auth/default_authorizer.cc

```diff
@@ -303,6 +311,55 @@ future<> default_authorizer::revoke_all(const resource& resource) {
     });
 }
 
+future<> default_authorizer::revoke_all(const resource& resource) {
+    if (legacy_mode(_qp)) {
+        co_return co_await revoke_all_legacy(resource);
+    }
+    auto name = resource.name();
+    try {
+        // guard has to be taken before we execute select so that
+        // operation is linearizable with other auth statements
+        // in other words delete and select below operate on the same data
+        auto group0_guard = co_await _group0_client.start_operation(&_as);
+        auto timestamp = group0_guard.write_timestamp();
+        auto gen = [this, name, timestamp] () -> mutations_generator {
+            static const sstring query = format("SELECT {} FROM {}.{} WHERE {} = ? ALLOW FILTERING",
+                ROLE_NAME,
+                _auth_ks_name,
+                PERMISSIONS_CF,
+                RESOURCE_NAME);
+            auto res = co_await _qp.execute_internal(
+                query,
+                db::consistency_level::LOCAL_ONE,
+                {name},
+                cql3::query_processor::cache_internal::no);
+            for (const auto& r : *res) {
+                 static const sstring query = format("DELETE FROM {}.{} WHERE {} = ? AND {} = ?",
+                    _auth_ks_name,
+                    PERMISSIONS_CF,
+                    ROLE_NAME,
+                    RESOURCE_NAME);
+                auto muts = co_await _qp.get_mutations_internal(
```

Comment:
Each of these can be called under a different guard, they need to take that guard's timestamp.

And the read result on which we base our decisions to create these mutations may no longer be valid due to concurrent operations which may happen due to our guard dropping and retaking, but I understand this is the FIXME part of the PR? This function doesn't have a FIXME though. We also need an issue once PR is merged

Reply:
yes, it should be possible to solve one case by adding guard as parameter, but there are two unsolved cases with current group0 API and this applies not only to auth, I believe most features on top of it will face those problems:
- get_mutations_internal may return more than one mutation which are currently co_yield'ed in a loop so they share the same timestamp
- command_is_too_big_error in theory still can be thrown and I need to somehow progress with work because if I simply retry the whole statement (like on concurrent modification) it will most likely be thrown again.

I don't know how to reasonably solve those. I guess the fundamental problem is that as group0 user I want to announce command of any size on a single guard.

>but I understand this is the FIXME part of the PR 

I will create issues after this is merged, can't do this before that because it changes a lot during this review.

Reply:
> pointer

Or simply a boolean that says the data is ready.

Reply:
if your system stores small amounts of data typically do you want to design it like it'd handle lots of data, and what if it's put to the limit, should it break / break the whole environment / loosen the guarantees? 

Here I opted for the latter, perhaps adding the guard reference closes the gap even further and is good enough? I don't know, I had to decide something and I don't think re-engineering the whole auth implementation would be a better choice.

Reply:
We discussed this on a call, it should be possible to fix the problem on a follow-up, e.g. instead of trying to delete all roles immediately, commit an intent to delete them and have them deleted by a background process. No need to do this now.